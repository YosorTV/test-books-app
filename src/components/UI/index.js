export * from './Loader';
export * from './Book';
export * from './Layout';