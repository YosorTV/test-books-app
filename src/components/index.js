export * from './BooksList';
export * from './BookDetail';