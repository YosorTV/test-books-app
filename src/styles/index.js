import styled from 'styled-components';
import { Card, Paper } from '@material-ui/core';
export * from './theme';

export const Load = styled.div`
  background: linear-gradient(to bottom, #ffffff, #f1f1f1 50%, #d3d3d3 75%, #d3d3d3 75%);
  min-height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`
export const Nav = styled.nav`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1rem;
  position: relative;
    svg{
      font-size: 3rem;
      color:#534646;
    }
    .return{
      position: absolute;
      left: 0.5rem;
      svg{
        font-size: 2rem;
        color: #303030;
      }
    }
`;
export const Main = styled.main`
  padding:4rem;
  display: flex;
  min-height: 100vh;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  ${({ theme }) => theme.breakpoints.down("md")} {
    padding:2rem;
  }
  ${({ theme }) => theme.breakpoints.down("xs")} {
    padding:.5rem;
  }
`;
export const StyledCard = styled(Card)`
  margin: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  .card-image{
    min-height: 25vh;
    width: 100%;
    object-fit: cover; 
  }
`
export const StyledCardDetail = styled(Paper)`
  width: 45vw;
  ${({ theme }) => theme.breakpoints.down("md")} {
    width: 60vw;
  }
  ${({ theme }) => theme.breakpoints.down("sm")} {
    width: 75vw;
  }
  ${({ theme }) => theme.breakpoints.down("xs")} {
    width: 90vw;
  }
  .book-info{
    width: 100%;
    display: flex;
    flex-direction: column;
    &__media{
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
    &__content{
      background: #f3f3f3;
      color:#303030;
      padding:1rem 2rem;
      &-description{
        display: flex;
        justify-content: space-between;
        align-items: baseline;
        &__title{
          font-weight: bold;
          font-size: 1.4rem;
        }
        &__author{
          font-style: italic;
          font-size: 1rem;
        }
        ${({ theme }) => theme.breakpoints.down("sm")} {
          flex-direction: column;
          &__title{
            font-size: 1rem;
          }
          &__author{
            padding-top: .5rem;
          }
        }
      }
      &-copy{
        display: flex;
        justify-content: space-between;
        padding: 1rem 0rem;
        &__about{
          flex: 1 1 100%;
          padding-right: 1rem;
        }
        &__isbn, &__action{
          span{
            font-weight: bold;
          }
          font-style: italic;
          font-size: 1rem;
        }
        ${({ theme }) => theme.breakpoints.down("sm")} {
          flex-direction: column;
            &__isbn, &__action{
            padding:1rem 0 0 0;
            flex: 1;
          }
        }
      }
    }
  }
`;