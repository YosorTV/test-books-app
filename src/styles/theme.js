import { createTheme } from "@material-ui/core";

export const theme = createTheme({
  palette: {
    type:'dark',
    primary:{
      main:"#f3f3f3",
      secondary:"#f6f19b"
    }
  },
  overrides:{
    MuiCssBaseline: {
      '@global': {
        "*":{
          boxSizing:"border-box",
          margin:0,
          padding:0,
          listStyle:"none"
        }
      }
    }
  }
})