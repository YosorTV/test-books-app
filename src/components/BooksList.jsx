// Core
import React, { useContext } from 'react'
// MUI
import { Grid } from '@material-ui/core';

import { Context } from '../HOC'
import { Book } from './UI';

export const BooksList = () => {
  const books = useContext(Context);
  const listOfBooksJSX = books?.map((book) => <Book key={book.id} {...book} />);
  return (
    <Grid container component="ul">
      {listOfBooksJSX}
    </Grid>
  )
}
