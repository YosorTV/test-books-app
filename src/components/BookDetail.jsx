import React, { Fragment, useContext, useEffect, useState } from 'react'
import { useParams } from 'react-router';
import { Context } from '../HOC';
import { Box, Typography } from '@material-ui/core';
import { StyledCardDetail } from '../styles';
import { booksBySameAuthor } from '../helpers';
import { Link } from 'react-router-dom';

export const BookDetail = () => {
  const books = useContext(Context);
  const { bookId } = useParams();
  const [bySameAuthor, setBySameAuthor] = useState(null);
  const [content, setContent] = useState(null);

  // setting detailed book information
  useEffect(() => {
    const data = books?.find(({ id }) => id === bookId);
    setContent(data);
  },[bookId, books])
  // setting list of books by same author
  useEffect(() => {
    const data = booksBySameAuthor(books, content?.author, bookId);
    setBySameAuthor(data);
  },[books, content?.author, bookId]);

  // render list of books by same author
  const listOfBooksBySameAuthor = bySameAuthor?.map(({id, title}) => 
    <Box component="ul" key={id}>
      <Box component="li">
        <Box component={Link} to={`/all-books/${id}`}> {title} </Box>
      </Box>
    </Box>)

  // render another books section
  const anotherBooksJSX = bySameAuthor && 
  <>
    <span>Another books by author:</span> 
    {listOfBooksBySameAuthor}
  </>

  return (
    <StyledCardDetail component="article">
    {content?.author && 
      <Box component="figure" className="book-info">
        <img src={content?.cover} title={`${content?.title} by - ${content?.author}`} alt="cover" className="book-info__media"/>
        <Box component="figcaption" className="book-info__content">
          <Box className="book-info__content-description">
            <Typography component="p" className="book-info__content-description__title">
                {content?.title}
              </Typography>
              <Typography component="p" className="book-info__content-description__author">
                {content?.author}
              </Typography>
          </Box>
          <Box className="book-info__content-copy">
          {content?.description && 
            <Typography component="p" className="book-info__content-copy__about">
              {content?.description}
            </Typography>}
            <Box component="div" display="flex" flexDirection="column">
              <Typography component="p" className="book-info__content-copy__isbn">
                <span>Serial number:</span> {content?.isbn}
              </Typography>
              <Typography component="div" className="book-info__content-copy__action">
                {anotherBooksJSX} 
              </Typography>
            </Box>
          </Box>
        </Box>
      </Box>
    }
    </StyledCardDetail>
  )
}
