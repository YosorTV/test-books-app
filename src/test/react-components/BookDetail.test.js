// Core
import Enzyme, {shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { render, unmountComponentAtNode } from 'react-dom';
import { BookDetail } from '../../components';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../styles';
import toJSON from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });

const init = () => {
  const wrapper = shallow(<BookDetail />);
  return { wrapper };
};

describe('BookDetail component', () => {
  test('BookDetail should render without crashings', () => {
    const section = document.createElement('section');
      render(
        <ThemeProvider theme={theme}>
          <BookDetail />
        </ThemeProvider>,
      section);
      unmountComponentAtNode(section);
  })
  
  test('BookDetail snapshot', () => {
    const { wrapper } = init();
    expect(toJSON(wrapper)).toMatchSnapshot();
  })
})