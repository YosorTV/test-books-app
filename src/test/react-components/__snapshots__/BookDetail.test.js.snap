// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetail component BookDetail snapshot 1`] = `
<Styled(WithStyles(ForwardRef(Paper)))
  component="article"
/>
`;
